src/rgw/rgw_directory.cc

```diff
@@ -0,0 +1,332 @@
+#include <errno.h>
+#include <cpp_redis/cpp_redis>
+#include "rgw_directory.h"
+#include <string>
+#include <iostream>
+#include <sstream>
+#include <algorithm>
+#include <vector>
+#include <list>
+
+void RGWBlockDirectory::findClient(string key, cpp_redis::client *client, int port) {
+  if (client->is_connected()) 
+    return;
+  
+  client->connect("127.0.0.1", port, nullptr);
+  
+  if (!client->is_connected())
+    exit(1);
+}
+
+string RGWBlockDirectory::buildIndex(cache_block *ptr) {
+  return ptr->c_obj.obj_name;
+}
+
+int RGWBlockDirectory::existKey(string key,cpp_redis::client *client) {
+  int result = 0;
+  vector<string> keys;
+  keys.push_back(key);
+  
+  if (!client->is_connected()) {
+    return result;
+  }
+
+  try {
+    client->exists(keys, [&result](cpp_redis::reply &reply) {
+      if (reply.is_integer()) {
+        result = reply.as_integer();
+      }
+    });
+    
+    client->sync_commit(std::chrono::milliseconds(1000));
+  }
+  catch(exception &e) {}
+
+  return result;
+}
+
+int RGWBlockDirectory::setValue(cache_block *ptr) {
+  //creating the index based on bucket_name, obj_name, and chunk_id
+  string key = buildIndex(ptr);
+
+  if (!client.is_connected()) { 
+    findClient(key, &client, 6379);
+  }
+
+  string result;
+  int exist = 0;
+  vector<string> keys;
+  keys.push_back(key);
+
+  try {
+    client.exists(keys, [&exist](cpp_redis::reply &reply) {
+      if (reply.is_integer()) {
+        exist = reply.as_integer();
+      }
+    });
+    
+    client.sync_commit(std::chrono::milliseconds(1000));
+  }
+  catch(exception &e) {
+    exist = 0;
+  }
+
+  if (!exist) {
+    vector<pair<string, string>> list;
+    
+    //creating a list of key's properties
+    list.push_back(make_pair("key", key));
+    list.push_back(make_pair("size", to_string(ptr->size_in_bytes)));
```

Comment:
Can we get away with only having one of these fields besides `key` that we add and remove the rest? Maybe just keep `obj_name` and `size`? I don't see any use for `bucket_name`, `hosts_list`, or even `size`. The change would have to be applied to the `cache_block` structs as well.

Reply:
Seg faults occur when `bucket_name`, `size`, and `hosts_list` are removed. As a result, I will keep them in the struct.